File: apps/ios/Ameme/Shared/LocalRecoveryPointManager.swift
import Foundation

public enum LocalRecoveryPointAvailability: String, Equatable, Sendable {
    case none
    case verified
    case unavailable
}

public struct LocalRecoveryPointStatus: Equatable, Sendable {
    public let availability: LocalRecoveryPointAvailability
    public let backupID: String?
    public let createdAt: Date?
    public let snapshotBytes: Int64?
    public let verifiedAt: Date?
    public let lastActivatedAt: Date?
    public let cleanupPending: Bool

    public static let none = LocalRecoveryPointStatus(
        availability: .none,
        backupID: nil,
        createdAt: nil,
        snapshotBytes: nil,
        verifiedAt: nil,
        lastActivatedAt: nil,
        cleanupPending: false
    )

    public static let unavailable = LocalRecoveryPointStatus(
        availability: .unavailable,
        backupID: nil,
        createdAt: nil,
        snapshotBytes: nil,
        verifiedAt: nil,
        lastActivatedAt: nil,
        cleanupPending: false
    )
}

/// Owns one bounded same-install recovery point outside the live encrypted-store root.
///
/// A point is reported as verified only after the artifact has been authenticated and restored
/// into a new isolated candidate directory under the current device-bound key. It never exports
/// key material and cannot make an uninstall, device-loss, or cross-device recovery claim.
@MainActor
public final class LocalRecoveryPointManager {
    private struct ActivationRecord: Codable {
        static let currentSchemaVersion = 1

        let schemaVersion: Int
        let backupID: String
        let activatedAtEpochMilliseconds: Int64
        let cleanupPending: Bool
        let productionRecoveryClaim: Bool
    }

    private let fileManager: FileManager
    private let rootDirectory: URL
    private var preparedCandidate: LocalRecoveryCandidate?
    private var preparedStatus: LocalRecoveryPointStatus = .none

    private var currentDirectory: URL {
        rootDirectory.appendingPathComponent("current", isDirectory: true)
    }

    private var previousDirectory: URL {
        rootDirectory.appendingPathComponent("previous", isDirectory: true)
    }

    private var candidateDirectory: URL {
        rootDirectory.appendingPathComponent("candidate", isDirectory: true)
    }

    private var activationRecordURL: URL {
        rootDirectory.appendingPathComponent("last-activation-v1.json")
    }

    public init(
        rootDirectory: URL,
        fileManager: FileManager = .default
    ) {
        self.rootDirectory = rootDirectory.standardizedFileURL
        self.fileManager = fileManager
    }

    @discardableResult
    public func refresh(
        using store: LocalMemoryStore,
        verifiedAt: Date = .now
    ) -> LocalRecoveryPointStatus {
        preparedCandidate = nil
        preparedStatus = .none
        do {
            try ensureSafeRoot()
            try removeCoreStagingDirectories()
            let sources = try recoverySources()
            guard !sources.isEmpty else {
                try removeCandidateIfPresent()
                return .none
            }

            var lastError: Error?
            for source in sources {
                do {
                    let prepared = try prepare(
                        source: source,
                        using: store,
                        verifiedAt: verifiedAt
                    )
                    if source.standardizedFileURL != currentDirectory.standardizedFileURL {
                        try replaceCurrent(with: source)
                    }
                    try removeSupersededSlots()
                    preparedCandidate = prepared.candidate
                    preparedStatus = prepared.status
                    return prepared.status
                } catch {
                    lastError = error
                    try? removeCandidateIfPresent()
                }
            }
            if let lastError {
                throw lastError
            }
            return .unavailable
        } catch {
            preparedCandidate = nil
            preparedStatus = .unavailable
            return .unavailable
        }
    }

    @discardableResult
    public func create(
        using store: LocalMemoryStore,
        createdAt: Date = .now
    ) throws -> LocalRecoveryPointStatus {
        try ensureSafeRoot()
        try recoverInterruptedRotation()
        try removeTransientSlots()
        let next = rootDirectory.appendingPathComponent(
            ".next-\(UUID().uuidString.lowercased())",
            isDirectory: true
        )
        do {
            _ = try store.createLocalRecoveryBackup(at: next, createdAt: createdAt)
            _ = try store.verifyLocalRecoveryBackup(at: next)
            try removeCandidateIfPresent()
            _ = try store.restoreLocalRecoveryCandidate(
                from: next,
                to: candidateDirectory
            )
            try rotateVerifiedNextIntoCurrent(next)
            let status = refresh(using: store)
            guard status.availability == .verified else {
                throw LocalBackupError.invalidArtifact
            }
            return status
        } catch {
            try? recoverInterruptedRotation()
            try? removeOwnedTree(next)
            _ = refresh(using: store)
            throw error
        }
    }

    @discardableResult
    public func activate(
        using store: LocalMemoryStore,
        confirmedAt: Date = .now
    ) throws -> LocalRecoveryPointStatus {
        guard preparedStatus.availability == .verified,
              let candidate = preparedCandidate,
              preparedStatus.backupID == candidate.sourceManifest.backupID else {
            throw LocalBackupError.invalidArtifact
        }
        let authorization = LocalRecoveryActivationAuthorization(
            confirmationID:
                LocalRecoveryActivationAuthorization.confirmationPrefix +
                UUID().uuidString.lowercased(),
            candidateBackupID: candidate.sourceManifest.backupID,
            confirmedAt: confirmedAt,
            expiresAt: confirmedAt.addingTimeInterval(10 * 60)
        )
        let receipt = try store.activateLocalRecoveryCandidate(
            candidate,
            authorization: authorization,
            activatedAt: confirmedAt
        )
        guard !receipt.productionRecoveryClaim,
              receipt.candidateBackupID == candidate.sourceManifest.backupID else {
            throw LocalBackupError.activationAuthorizationInvalid
        }
        try writeActivationRecord(receipt)
        let status = refresh(using: store)
        guard status.availability == .verified,
              status.lastActivatedAt != nil else {
            throw LocalBackupError.invalidArtifact
        }
        return status
    }

    public func clear() throws {
        preparedCandidate = nil
        preparedStatus = .none
        guard fileManager.fileExists(atPath: rootDirectory.path) else { return }
        try removeOwnedTree(rootDirectory, allowRoot: true)
    }

    private func prepare(
        source: URL,
        using store: LocalMemoryStore,
        verifiedAt: Date
    ) throws -> (candidate: LocalRecoveryCandidate, status: LocalRecoveryPointStatus) {
        try requireOwnedSlot(source)
        let manifest = try store.verifyLocalRecoveryBackup(at: source)
        try removeCandidateIfPresent()
        let candidate = try store.restoreLocalRecoveryCandidate(
            from: source,
            to: candidateDirectory
        )
        guard candidate.sourceManifest == manifest,
              !candidate.productionRecoveryClaim,
              !manifest.productionRecoveryClaim else {
            throw LocalBackupError.invalidArtifact
        }
        let record = readActivationRecord(matching: manifest.backupID)
        let status = LocalRecoveryPointStatus(
            availability: .verified,
            backupID: manifest.backupID,
            createdAt: Self.parseISO8601(manifest.createdAt),
            snapshotBytes: manifest.eventsCiphertextSize +
                manifest.media.reduce(0) { $0 + $1.size },
            verifiedAt: verifiedAt,
            lastActivatedAt: record.map {
                Date(
                    timeIntervalSince1970:
                        Double($0.activatedAtEpochMilliseconds) / 1_000
                )
            },
            cleanupPending: record?.cleanupPending ?? false
        )
        return (candidate, status)
    }

    private func recoverySources() throws -> [URL] {
        try recoverInterruptedRotation()
        let currentExists = fileManager.fileExists(atPath: currentDirectory.path)
        let previousExists = fileManager.fileExists(atPath: previousDirectory.path)
        let next = try transientDirectories(prefix: ".next-")
        guard next.count <= 1 else {
            throw LocalBackupError.invalidArtifact
        }
        if currentExists {
            return [currentDirectory] + next + (previousExists ? [previousDirectory] : [])
        }
        if next.count == 1 {
            return next + (previousExists ? [previousDirectory] : [])
        }
        if next.count > 1 {
            throw LocalBackupError.invalidArtifact
        }
        return previousExists ? [previousDirectory] : []
    }

    private func recoverInterruptedRotation() throws {
        guard fileManager.fileExists(atPath: rootDirectory.path) else { return }
        try validateRootEntries()
        try removeCoreStagingDirectories()
        if !fileManager.fileExists(atPath: currentDirectory.path) {
            let next = try transientDirectories(prefix: ".next-")
            if next.count == 1 {
                try fileManager.moveItem(at: next[0], to: currentDirectory)
            } else if next.isEmpty,
                      fileManager.fileExists(atPath: previousDirectory.path) {
                try fileManager.moveItem(at: previousDirectory, to: currentDirectory)
            } else if next.count > 1 {
                throw LocalBackupError.invalidArtifact
            }
        }
    }

    private func replaceCurrent(with source: URL) throws {
        try requireOwnedSlot(source)
        if source.standardizedFileURL == currentDirectory.standardizedFileURL {
            return
        }
        if fileManager.fileExists(atPath: currentDirectory.path) {
            try removeOwnedTree(currentDirectory)
        }
        try fileManager.moveItem(at: source, to: currentDirectory)
    }

    private func rotateVerifiedNextIntoCurrent(_ next: URL) throws {
        try requireOwnedSlot(next)
        if fileManager.fileExists(atPath: previousDirectory.path) {
            try removeOwnedTree(previousDirectory)
        }
        if fileManager.fileExists(atPath: currentDirectory.path) {
            try fileManager.moveItem(at: currentDirectory, to: previousDirectory)
        }
        try fileManager.moveItem(at: next, to: currentDirectory)
        if fileManager.fileExists(atPath: previousDirectory.path) {
            try removeOwnedTree(previousDirectory)
        }
        try removeActivationRecordIfPresent()
    }

    private func removeSupersededSlots() throws {
        if fileManager.fileExists(atPath: previousDirectory.path) {
            try removeOwnedTree(previousDirectory)
        }
        try removeTransientSlots()
        try removeCoreStagingDirectories()
    }

    private func removeTransientSlots() throws {
        for url in try transientDirectories(prefix: ".next-") {
            try removeOwnedTree(url)
        }
        let temporaryRecords = try fileManager.contentsOfDirectory(
            at: rootDirectory,
            includingPropertiesForKeys: [.isRegularFileKey, .isSymbolicLinkKey]
        ).filter { isActivationTemporaryName($0.lastPathComponent) }
        for temporary in temporaryRecords {
            guard Self.isRegularNonSymbolicFile(
                temporary,
                fileManager: fileManager
            ) else {
                throw LocalBackupError.invalidArtifact
            }
            try fileManager.removeItem(at: temporary)
        }
    }

    private func removeCoreStagingDirectories() throws {
        guard fileManager.fileExists(atPath: rootDirectory.path) else { return }
        let staging = try fileManager.contentsOfDirectory(
            at: rootDirectory,
            includingPropertiesForKeys: [.isDirectoryKey, .isSymbolicLinkKey]
        ).filter { isCoreStagingName($0.lastPathComponent) }
        for directory in staging {
            try removeOwnedTree(directory)
        }
    }

    private func removeCandidateIfPresent() throws {
        if fileManager.fileExists(atPath: candidateDirectory.path) {
            try removeOwnedTree(candidateDirectory)
        }
    }

    private func removeActivationRecordIfPresent() throws {
        if fileManager.fileExists(atPath: activationRecordURL.path) {
            guard Self.isRegularNonSymbolicFile(
                activationRecordURL,
                fileManager: fileManager
            ) else {
                throw LocalBackupError.invalidArtifact
            }
            try fileManager.removeItem(at: activationRecordURL)
        }
    }

    private func ensureSafeRoot() throws {
        if !fileManager.fileExists(atPath: rootDirectory.path) {
            try fileManager.createDirectory(
                at: rootDirectory,
                withIntermediateDirectories: true
            )
        }
        guard let values = try? rootDirectory.resourceValues(
            forKeys: [.isDirectoryKey, .isSymbolicLinkKey]
        ),
        values.isDirectory == true,
        values.isSymbolicLink != true else {
            throw LocalBackupError.invalidArtifact
        }
        try validateRootEntries()
    }

    private func validateRootEntries() throws {
        let urls = try fileManager.contentsOfDirectory(
            at: rootDirectory,
            includingPropertiesForKeys: [.isDirectoryKey, .isRegularFileKey, .isSymbolicLinkKey]
        )
        for url in urls {
            let name = url.lastPathComponent
            let allowed = name == "current" ||
                name == "previous" ||
                name == "candidate" ||
                name == activationRecordURL.lastPathComponent ||
                isTransientDirectoryName(name, prefix: ".next-") ||
                isCoreStagingName(name) ||
                isActivationTemporaryName(name)
            guard allowed else {
                throw LocalBackupError.invalidArtifact
            }
            let values = try url.resourceValues(
                forKeys: [.isDirectoryKey, .isRegularFileKey, .isSymbolicLinkKey]
            )
            guard values.isSymbolicLink != true else {
                throw LocalBackupError.invalidArtifact
            }
        }
    }

    private func transientDirectories(prefix: String) throws -> [URL] {
        let urls = try fileManager.contentsOfDirectory(
            at: rootDirectory,
            includingPropertiesForKeys: [.isDirectoryKey, .isSymbolicLinkKey]
        ).filter { $0.lastPathComponent.hasPrefix(prefix) }
        for url in urls {
            let values = try url.resourceValues(
                forKeys: [.isDirectoryKey, .isSymbolicLinkKey]
            )
            guard values.isDirectory == true,
                  values.isSymbolicLink != true,
                  UUID(
                    uuidString: String(url.lastPathComponent.dropFirst(prefix.count))
                  ) != nil else {
                throw LocalBackupError.invalidArtifact
            }
        }
        return urls.sorted { $0.lastPathComponent < $1.lastPathComponent }
    }

    private func requireOwnedSlot(_ url: URL) throws {
        let normalized = url.standardizedFileURL
        guard normalized.deletingLastPathComponent() == rootDirectory,
              normalized != rootDirectory else {
            throw LocalBackupError.invalidArtifact
        }
        let name = normalized.lastPathComponent
        guard name == "current" ||
                name == "previous" ||
                name == "candidate" ||
                isTransientDirectoryName(name, prefix: ".next-") ||
                isCoreStagingName(name) else {
            throw LocalBackupError.invalidArtifact
        }
    }

    private func removeOwnedTree(_ url: URL, allowRoot: Bool = false) throws {
        let normalized = url.standardizedFileURL
        if allowRoot {
            guard normalized == rootDirectory else {
                throw LocalBackupError.invalidArtifact
            }
        } else {
            try requireOwnedSlot(normalized)
        }
        guard fileManager.fileExists(atPath: normalized.path),
              let enumerator = fileManager.enumerator(
                at: normalized,
                includingPropertiesForKeys: [.isSymbolicLinkKey],
                options: []
              ) else {
            throw LocalBackupError.invalidArtifact
        }
        var descendants: [URL] = []
        for case let child as URL in enumerator {
            let values = try child.resourceValues(forKeys: [.isSymbolicLinkKey])
            guard values.isSymbolicLink != true,
                  child.standardizedFileURL.path.hasPrefix(
                    normalized.path + "/"
                  ) else {
                throw LocalBackupError.invalidArtifact
            }
            descendants.append(child)
        }
        for child in descendants.sorted(by: { $0.path.count > $1.path.count }) {
            try fileManager.removeItem(at: child)
        }
        try fileManager.removeItem(at: normalized)
    }

    private func writeActivationRecord(
        _ receipt: LocalRecoveryActivationReceipt
    ) throws {
        let record = ActivationRecord(
            schemaVersion: ActivationRecord.currentSchemaVersion,
            backupID: receipt.candidateBackupID,
            activatedAtEpochMilliseconds: Int64(
                (receipt.activatedAt.timeIntervalSince1970 * 1_000).rounded()
            ),
            cleanupPending: receipt.cleanupPending,
            productionRecoveryClaim: false
        )
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.sortedKeys]
        let data = try encoder.encode(record)
        guard data.count <= 4 * 1_024 else {
            throw LocalBackupError.invalidArtifact
        }
        let temporary = rootDirectory.appendingPathComponent(
            ".last-activation-v1.json.tmp-\(UUID().uuidString.lowercased())"
        )
        do {
            try data.write(to: temporary, options: [.atomic])
            if fileManager.fileExists(atPath: activationRecordURL.path) {
                try removeActivationRecordIfPresent()
            }
            try fileManager.moveItem(at: temporary, to: activationRecordURL)
        } catch {
            try? fileManager.removeItem(at: temporary)
            throw error
        }
    }

    private func readActivationRecord(
        matching backupID: String
    ) -> ActivationRecord? {
        guard Self.isRegularNonSymbolicFile(
            activationRecordURL,
            fileManager: fileManager
        ),
        let attributes = try? fileManager.attributesOfItem(
            atPath: activationRecordURL.path
        ),
        let size = attributes[.size] as? NSNumber,
        size.intValue > 0,
        size.intValue <= 4 * 1_024,
        let data = try? Data(contentsOf: activationRecordURL),
        let record = try? JSONDecoder().decode(ActivationRecord.self, from: data),
        record.schemaVersion == ActivationRecord.currentSchemaVersion,
        record.backupID == backupID,
        record.backupID.hasPrefix("backup_"),
        UUID(
            uuidString: String(record.backupID.dropFirst("backup_".count))
        ) != nil,
        record.activatedAtEpochMilliseconds >= 0,
        !record.productionRecoveryClaim else {
            return nil
        }
        return record
    }

    private static func isRegularNonSymbolicFile(
        _ url: URL,
        fileManager: FileManager
    ) -> Bool {
        guard fileManager.fileExists(atPath: url.path),
              let values = try? url.resourceValues(
                forKeys: [.isRegularFileKey, .isSymbolicLinkKey]
              ) else {
            return false
        }
        return values.isRegularFile == true && values.isSymbolicLink != true
    }

    private static func parseISO8601(_ value: String) -> Date? {
        let fractional = ISO8601DateFormatter()
        fractional.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return fractional.date(from: value) ?? ISO8601DateFormatter().date(from: value)
    }

    private func isTransientDirectoryName(
        _ name: String,
        prefix: String
    ) -> Bool {
        name.hasPrefix(prefix) &&
            UUID(uuidString: String(name.dropFirst(prefix.count))) != nil
    }

    private func isCoreStagingName(_ name: String) -> Bool {
        let candidatePrefix = ".candidate.staging-"
        if name.hasPrefix(candidatePrefix) {
            return UUID(
                uuidString: String(name.dropFirst(candidatePrefix.count))
            ) != nil
        }
        let nextPrefix = "..next-"
        guard name.hasPrefix(nextPrefix) else { return false }
        let fields = String(name.dropFirst(nextPrefix.count))
            .components(separatedBy: ".staging-")
        return fields.count == 2 &&
            fields.allSatisfy { UUID(uuidString: $0) != nil }
    }

    private func isActivationTemporaryName(_ name: String) -> Bool {
        let prefix = ".last-activation-v1.json.tmp-"
        return name.hasPrefix(prefix) &&
            UUID(uuidString: String(name.dropFirst(prefix.count))) != nil
    }
}

File: apps/android/app/src/main/java/com/ameme/android/data/local/LocalRecoveryPointManager.kt
package com.ameme.android.data.local

import com.ameme.android.data.DeletionWatermark
import com.ameme.android.data.RecoveryActivationAuthorization
import com.ameme.android.data.RecoveryActivationReceipt
import com.ameme.android.data.RecoveryBackupRepository
import com.ameme.android.data.RecoveryCandidate
import java.io.File
import java.io.RandomAccessFile
import java.nio.charset.StandardCharsets
import java.nio.file.Files
import java.nio.file.StandardCopyOption
import java.time.Instant
import java.util.UUID

enum class LocalRecoveryPointAvailability {
    None,
    Verified,
    Unavailable,
}

data class LocalRecoveryPointStatus(
    val availability: LocalRecoveryPointAvailability,
    val backupId: String? = null,
    val createdAt: Instant? = null,
    val snapshotBytes: Long? = null,
    val verifiedAt: Instant? = null,
    val lastActivatedAt: Instant? = null,
    val cleanupPending: Boolean = false,
) {
    companion object {
        val None = LocalRecoveryPointStatus(LocalRecoveryPointAvailability.None)
        val Unavailable = LocalRecoveryPointStatus(LocalRecoveryPointAvailability.Unavailable)
    }
}

data class LocalRecoveryActivationPlan(
    val candidate: RecoveryCandidate,
    val authorization: RecoveryActivationAuthorization,
    val authoritativeWatermarks: List<DeletionWatermark>,
)

/**
 * Owns one bounded same-install recovery point outside the live SQLCipher directory.
 *
 * A point is healthy only after the authenticated artifact has been restored into a fresh,
 * isolated candidate under the current device-bound key and authoritative deletion watermarks.
 * This manager never exports key material and cannot claim uninstall, device-loss, or
 * cross-device recovery.
 */
class LocalRecoveryPointManager(
    rootDirectory: File,
) {
    private data class Prepared(
        val candidate: RecoveryCandidate,
        val status: LocalRecoveryPointStatus,
    )

    private data class ActivationRecord(
        val backupId: String,
        val activatedAt: Instant,
        val cleanupPending: Boolean,
    )

    private val rootDirectory = rootDirectory.absoluteFile
    private val currentDirectory = File(this.rootDirectory, CURRENT)
    private val previousDirectory = File(this.rootDirectory, PREVIOUS)
    private val candidateDirectory = File(this.rootDirectory, CANDIDATE)
    private val activationRecordFile = File(this.rootDirectory, ACTIVATION_RECORD)
    private var preparedCandidate: RecoveryCandidate? = null
    private var preparedStatus: LocalRecoveryPointStatus = LocalRecoveryPointStatus.None

    fun refresh(
        repository: RecoveryBackupRepository,
        verifiedAt: Instant = Instant.now(),
    ): LocalRecoveryPointStatus {
        preparedCandidate = null
        preparedStatus = LocalRecoveryPointStatus.None
        return runCatching {
            ensureSafeRoot()
            removeCoreStagingDirectories()
            val sources = recoverySources()
            if (sources.isEmpty()) {
                removeCandidateIfPresent()
                return@runCatching LocalRecoveryPointStatus.None
            }

            var lastError: Throwable? = null
            for (source in sources) {
                try {
                    val prepared = prepare(source, repository, verifiedAt)
                    if (source.canonicalFile != currentDirectory.canonicalFile) {
                        replaceCurrent(source)
                    }
                    removeSupersededSlots()
                    preparedCandidate = prepared.candidate
                    preparedStatus = prepared.status
                    return@runCatching prepared.status
                } catch (error: Throwable) {
                    lastError = error
                    runCatching(::removeCandidateIfPresent)
                }
            }
            throw lastError ?: IllegalArgumentException("Recovery point is unavailable")
        }.getOrElse {
            preparedCandidate = null
            preparedStatus = LocalRecoveryPointStatus.Unavailable
            LocalRecoveryPointStatus.Unavailable
        }
    }

    fun create(
        repository: RecoveryBackupRepository,
        createdAt: Instant = Instant.now(),
    ): LocalRecoveryPointStatus {
        ensureSafeRoot()
        recoverInterruptedRotation()
        removeTransientSlots()
        val next = File(rootDirectory, "$NEXT_PREFIX${UUID.randomUUID()}")
        try {
            repository.createLocalRecoveryBackup(next, createdAt)
            repository.verifyLocalRecoveryBackup(next)
            removeCandidateIfPresent()
            repository.restoreLocalRecoveryCandidate(next, candidateDirectory)
            rotateVerifiedNextIntoCurrent(next)
            return refresh(repository).also {
                check(it.availability == LocalRecoveryPointAvailability.Verified) {
                    "Published recovery point did not pass isolated restore verification"
                }
            }
        } catch (error: Throwable) {
            runCatching(::recoverInterruptedRotation).onFailure(error::addSuppressed)
            runCatching { if (next.exists()) removeOwnedTree(next) }
                .onFailure(error::addSuppressed)
            refresh(repository)
            throw error
        }
    }

    fun prepareActivation(
        repository: RecoveryBackupRepository,
        confirmedAt: Instant = Instant.now(),
    ): LocalRecoveryActivationPlan {
        val candidate = requireNotNull(preparedCandidate) {
            "No verified same-install recovery candidate is prepared"
        }
        require(
            preparedStatus.availability == LocalRecoveryPointAvailability.Verified &&
                preparedStatus.backupId == candidate.sourceManifest.backupId,
        ) { "Prepared recovery candidate does not match the verified recovery point" }
        return LocalRecoveryActivationPlan(
            candidate = candidate,
            authorization = RecoveryActivationAuthorization(
                confirmationId =
                    RecoveryActivationAuthorization.CONFIRMATION_PREFIX +
                        UUID.randomUUID(),
                candidateBackupId = candidate.sourceManifest.backupId,
                confirmedAt = confirmedAt,
                expiresAt = confirmedAt.plusSeconds(ACTIVATION_VALIDITY_SECONDS),
            ),
            authoritativeWatermarks = repository.deletionWatermarks(),
        )
    }

    fun recordSuccessfulActivation(receipt: RecoveryActivationReceipt): Boolean {
        val expectedBackupId = preparedStatus.backupId ?: return false
        if (
            receipt.productionRecoveryClaim ||
            receipt.candidateBackupId != expectedBackupId
        ) {
            return false
        }
        return runCatching {
            writeActivationRecord(receipt)
            true
        }.getOrDefault(false)
    }

    fun clear() {
        preparedCandidate = null
        preparedStatus = LocalRecoveryPointStatus.None
        if (rootDirectory.exists()) {
            removeOwnedTree(rootDirectory, allowRoot = true)
        }
    }

    private fun prepare(
        source: File,
        repository: RecoveryBackupRepository,
        verifiedAt: Instant,
    ): Prepared {
        requireOwnedSlot(source)
        val manifest = repository.verifyLocalRecoveryBackup(source)
        removeCandidateIfPresent()
        val candidate = repository.restoreLocalRecoveryCandidate(source, candidateDirectory)
        require(
            candidate.sourceManifest == manifest &&
                !candidate.productionRecoveryClaim &&
                !manifest.productionRecoveryClaim,
        ) { "Prepared recovery candidate is inconsistent" }
        val activation = readActivationRecord(manifest.backupId)
        return Prepared(
            candidate = candidate,
            status = LocalRecoveryPointStatus(
                availability = LocalRecoveryPointAvailability.Verified,
                backupId = manifest.backupId,
                createdAt = Instant.parse(manifest.createdAt),
                snapshotBytes = manifest.snapshotSize,
                verifiedAt = verifiedAt,
                lastActivatedAt = activation?.activatedAt,
                cleanupPending = activation?.cleanupPending ?: false,
            ),
        )
    }

    private fun recoverySources(): List<File> {
        recoverInterruptedRotation()
        val next = transientDirectories(NEXT_PREFIX)
        require(next.size <= 1) { "Multiple interrupted recovery points are ambiguous" }
        return buildList {
            if (currentDirectory.exists()) add(currentDirectory)
            addAll(next)
            if (previousDirectory.exists()) add(previousDirectory)
        }.distinctBy { it.canonicalPath }
    }

    private fun recoverInterruptedRotation() {
        if (!rootDirectory.exists()) return
        validateRootEntries()
        removeCoreStagingDirectories()
        if (!currentDirectory.exists()) {
            val next = transientDirectories(NEXT_PREFIX)
            when {
                next.size == 1 -> atomicMove(next.single(), currentDirectory)
                next.size > 1 -> error("Multiple interrupted recovery points are ambiguous")
                previousDirectory.exists() -> atomicMove(previousDirectory, currentDirectory)
            }
        }
    }

    private fun replaceCurrent(source: File) {
        requireOwnedSlot(source)
        if (source.canonicalFile == currentDirectory.canonicalFile) return
        if (currentDirectory.exists()) removeOwnedTree(currentDirectory)
        atomicMove(source, currentDirectory)
    }

    private fun rotateVerifiedNextIntoCurrent(next: File) {
        requireOwnedSlot(next)
        if (previousDirectory.exists()) removeOwnedTree(previousDirectory)
        if (currentDirectory.exists()) atomicMove(currentDirectory, previousDirectory)
        atomicMove(next, currentDirectory)
        if (previousDirectory.exists()) removeOwnedTree(previousDirectory)
        removeActivationRecordIfPresent()
    }

    private fun removeSupersededSlots() {
        if (previousDirectory.exists()) removeOwnedTree(previousDirectory)
        removeTransientSlots()
        removeCoreStagingDirectories()
    }

    private fun removeTransientSlots() {
        transientDirectories(NEXT_PREFIX).forEach(::removeOwnedTree)
        rootDirectory.listFiles()
            .orEmpty()
            .filter { it.name.startsWith(ACTIVATION_TEMP_PREFIX) }
            .forEach { temporary ->
                requireRegularNonSymbolicFile(temporary)
                Files.delete(temporary.toPath())
            }
    }

    private fun removeCoreStagingDirectories() {
        if (!rootDirectory.isDirectory) return
        rootDirectory.listFiles()
            .orEmpty()
            .filter { isCoreStagingName(it.name) }
            .forEach(::removeOwnedTree)
    }

    private fun removeCandidateIfPresent() {
        if (candidateDirectory.exists()) removeOwnedTree(candidateDirectory)
    }

    private fun removeActivationRecordIfPresent() {
        if (activationRecordFile.exists()) {
            requireRegularNonSymbolicFile(activationRecordFile)
            Files.delete(activationRecordFile.toPath())
        }
    }

    private fun ensureSafeRoot() {
        if (!rootDirectory.exists()) {
            check(rootDirectory.mkdirs()) { "Recovery point root could not be created" }
        }
        require(
            rootDirectory.isDirectory &&
                !Files.isSymbolicLink(rootDirectory.toPath()),
        ) { "Recovery point root is unsafe" }
        validateRootEntries()
    }

    private fun validateRootEntries() {
        rootDirectory.listFiles().orEmpty().forEach { child ->
            require(!Files.isSymbolicLink(child.toPath())) {
                "Recovery point root contains a symbolic link"
            }
            require(
                child.name in setOf(CURRENT, PREVIOUS, CANDIDATE, ACTIVATION_RECORD) ||
                    isTransientDirectoryName(child.name, NEXT_PREFIX) ||
                    isCoreStagingName(child.name) ||
                    isActivationTemporaryName(child.name),
            ) { "Recovery point root contains an unknown artifact" }
        }
    }

    private fun transientDirectories(prefix: String): List<File> =
        rootDirectory.listFiles()
            .orEmpty()
            .filter { it.name.startsWith(prefix) }
            .onEach { directory ->
                require(
                    directory.isDirectory &&
                        !Files.isSymbolicLink(directory.toPath()) &&
                        isTransientDirectoryName(directory.name, prefix),
                ) { "Recovery point transient directory is invalid" }
            }
            .sortedBy(File::getName)

    private fun requireOwnedSlot(file: File) {
        val normalized = file.canonicalFile
        require(normalized.parentFile == rootDirectory.canonicalFile && normalized != rootDirectory) {
            "Recovery point path escapes its private root"
        }
        require(
            normalized.name in setOf(CURRENT, PREVIOUS, CANDIDATE) ||
                isTransientDirectoryName(normalized.name, NEXT_PREFIX) ||
                isCoreStagingName(normalized.name),
        ) { "Recovery point slot name is invalid" }
    }

    private fun removeOwnedTree(file: File, allowRoot: Boolean = false) {
        val normalized = file.canonicalFile
        if (allowRoot) {
            require(normalized == rootDirectory.canonicalFile) {
                "Only the exact recovery point root may be cleared"
            }
        } else {
            requireOwnedSlot(normalized)
        }
        require(normalized.exists()) { "Recovery point path does not exist" }
        val paths = Files.walk(normalized.toPath()).use { stream ->
            stream.toList()
        }
        paths.forEach { path ->
            require(!Files.isSymbolicLink(path)) {
                "Recovery point tree contains a symbolic link"
            }
            require(
                path.toFile().canonicalPath == normalized.canonicalPath ||
                    path.toFile().canonicalPath.startsWith("${normalized.canonicalPath}${File.separator}"),
            ) { "Recovery point tree escapes its private slot" }
        }
        paths.sortedByDescending { it.nameCount }.forEach(Files::delete)
    }

    private fun writeActivationRecord(receipt: RecoveryActivationReceipt) {
        val fields = listOf(
            ACTIVATION_RECORD_VERSION,
            receipt.candidateBackupId,
            receipt.activatedAt.toEpochMilli().toString(),
            receipt.cleanupPending.toString(),
            false.toString(),
        )
        val payload = fields.joinToString("\n", postfix = "\n").toByteArray(StandardCharsets.UTF_8)
        require(payload.size <= MAX_ACTIVATION_RECORD_BYTES) {
            "Recovery activation record is too large"
        }
        val temporary = File(
            rootDirectory,
            "$ACTIVATION_TEMP_PREFIX${UUID.randomUUID()}",
        )
        try {
            RandomAccessFile(temporary, "rw").use { output ->
                output.setLength(0)
                output.write(payload)
                output.fd.sync()
            }
            if (activationRecordFile.exists()) removeActivationRecordIfPresent()
            atomicMove(temporary, activationRecordFile)
        } finally {
            if (temporary.exists()) {
                requireRegularNonSymbolicFile(temporary)
                Files.delete(temporary.toPath())
            }
        }
    }

    private fun readActivationRecord(expectedBackupId: String): ActivationRecord? {
        if (!activationRecordFile.exists()) return null
        return runCatching {
            requireRegularNonSymbolicFile(activationRecordFile)
            require(activationRecordFile.length() in 1..MAX_ACTIVATION_RECORD_BYTES.toLong()) {
                "Recovery activation record size is invalid"
            }
            val lines = activationRecordFile.readLines(StandardCharsets.UTF_8)
            require(
                lines.size == 5 &&
                    lines[0] == ACTIVATION_RECORD_VERSION &&
                    lines[1] == expectedBackupId &&
                    lines[1].startsWith("backup_") &&
                    runCatching {
                        UUID.fromString(lines[1].removePrefix("backup_"))
                    }.isSuccess &&
                    lines[4] == false.toString(),
            ) { "Recovery activation record is invalid" }
            val activatedAt = Instant.ofEpochMilli(lines[2].toLong())
            require(!activatedAt.isBefore(Instant.EPOCH)) {
                "Recovery activation timestamp is invalid"
            }
            val cleanupPending = when (lines[3]) {
                true.toString() -> true
                false.toString() -> false
                else -> error("Recovery activation cleanup state is invalid")
            }
            ActivationRecord(lines[1], activatedAt, cleanupPending)
        }.getOrNull()
    }

    private fun requireRegularNonSymbolicFile(file: File) {
        require(file.isFile && !Files.isSymbolicLink(file.toPath())) {
            "Recovery point metadata file is unsafe"
        }
    }

    private fun isTransientDirectoryName(name: String, prefix: String): Boolean =
        name.startsWith(prefix) &&
            runCatching { UUID.fromString(name.removePrefix(prefix)) }.isSuccess

    private fun isCoreStagingName(name: String): Boolean {
        if (name.startsWith(CANDIDATE_STAGING_PREFIX)) {
            return runCatching {
                UUID.fromString(name.removePrefix(CANDIDATE_STAGING_PREFIX))
            }.isSuccess
        }
        if (!name.startsWith("..next-")) return false
        val parts = name.removePrefix("..next-").split(".staging-", limit = 2)
        return parts.size == 2 &&
            parts.all { runCatching { UUID.fromString(it) }.isSuccess }
    }

    private fun isActivationTemporaryName(name: String): Boolean =
        name.startsWith(ACTIVATION_TEMP_PREFIX) &&
            runCatching {
                UUID.fromString(name.removePrefix(ACTIVATION_TEMP_PREFIX))
            }.isSuccess

    private fun atomicMove(source: File, destination: File) {
        Files.move(
            source.toPath(),
            destination.toPath(),
            StandardCopyOption.ATOMIC_MOVE,
        )
    }

    private companion object {
        const val CURRENT = "current"
        const val PREVIOUS = "previous"
        const val CANDIDATE = "candidate"
        const val NEXT_PREFIX = ".next-"
        const val CANDIDATE_STAGING_PREFIX = ".candidate.staging-"
        const val ACTIVATION_RECORD = "last-activation-v1.txt"
        const val ACTIVATION_TEMP_PREFIX = ".last-activation-v1.txt.tmp-"
        const val ACTIVATION_RECORD_VERSION = "ameme-local-recovery-activation-record-v1"
        const val MAX_ACTIVATION_RECORD_BYTES = 4 * 1024
        const val ACTIVATION_VALIDITY_SECONDS = 10L * 60
    }
}
